keps/sig-node/4580-deprecate-kubelet-runonce/README.md

```diff
@@ -0,0 +1,355 @@
+
+### Goals
+
+* Mark the `RunOnce` field in `KubeletConfiguration` and the `runonce` flag of kubelet as deprecated,  and finally remove the runonce flag.
+* Remove kubelet support for RunOnce mode.
+
+### Non-Goals
+
+Immediate removal: the deprecation and removal process will be gradual and feature gate to increase awareness among potential users.
+
+## Proposal
+
+The runonce mode of kubelet will exit the kubelet process after spawning pods from the local manifests or remote URL. It is suitable for scenarios where one-time tasks need to be run on the node, this proposal outlines plans to deprecate and remove runonce mode in kubelet.
+
+### Risks and Mitigations
+
+Some people may still rely on this feature, but podman addresses the same use case with more well-supported way, ref: https://docs.podman.io/en/latest/markdown/podman-kube.1.html. Affected users can migrate to *podman kube subcommand* on demand.
+
+## Design Details
+
+### KubeletConfiguration Change: KubeletConfiguration
+
+Mark the `RunOnce` field as deprecated.
+
+### kubelet flag Change
+
+make the `runonce` flag as deprecated, and remove it in GA version.
+
+### Implement warning logging for runonce mode usage
+
+Starting in 1.31, during kubelet startup, if running in runonce mode, the kubelet will log a warning message, for example:
+
+```
+klog.Warning("RunOnce mode has been deprecated, and will be removed in a future release")
+```
+
+### Introduction LegacyNodeRunOnceMode feature gate
+
+With the introduction of the `LegacyNodeRunOnceMode` feature gate, Kubernetes aims to guide users through the deprecated runonce mode. When this feature gate is enabled, kubelet will refuse to start when the runonce flag is set.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- N/A
+
+##### Integration tests
+
+- N/A
+
+##### e2e tests
+
+- N/A
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature gate `LegacyNodeRunOnceMode` is introduced, is disabled by default. Enabling this feature gate will fail the kubelet on startup with runonce mode enable.
+- Mark the `RunOnce` field in KubeletConfiguration as deprecated.
+
+#### Beta
+
+- `LegacyNodeRunOnceMode` feature gate is enabled by default.
+- Failed when starting kubelet in runonce mode.
+
+#### GA
+
+- Remove `LegacyNodeRunOnceMode`  feature gate.
+- Comment the `RunOnce` field in KubeletConfiguration as 'no longer has any effect', and remove the kubelet's runonce flag.
+- Remove kubelet runonce mode.
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
+
+- N/A
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: LegacyNodeRunOnceMode
```

Comment:
See my other comment about feature gate name.

Reply:
We might need a different template for this kind of change.